File: src/Minkowski/minkowski.cpp
#include<iostream>
#include<math.h>
#include<random>
#include<chrono>
#include<ctime>
#include"matplotlibcpp.h"
# define pi 4*atan(1)
# define c 1
# define k 1
# define m 1
# define h 2*pi

namespace plt = matplotlibcpp;

unsigned t1, t2;
double dx = 0.8; // tamaño del paso para las derivadas.
int Tmaxima = 100; 

// Hamiltoniano en relatividad especial.
double H (double p_x, double p_y, double p_z) {
	return c*sqrt(pow(m*c,2) + pow(p_x,2) + pow(p_y,2) + pow(p_z,2))-m*pow(c,2);
}

// funcion de particion de una partícula.
double z (double V, double T) {
	unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
	std::mt19937 gen(seed);
	std::exponential_distribution<double> dist(1/(k*T));
	int N = 20000;
	int M = 100; 
	double s1 = 0;
	for (int i = 0; i < M; i++) {
		double s2 = 0;
		for (int j = 0; j < N; j++) {
			double p_x = dist(gen);
			double p_y = dist(gen);
			double p_z = dist(gen);
			s2 += exp(-1/(k*T)*H(p_x,p_y, p_z))/(pow(1/(k*T),3)*exp(-(1/(k*T))*(p_x + p_y + p_z)));
		}
		double integral = (8*s2)/N;
		s1 += integral;
	}
	double promedio = s1/M;
	return V*promedio/pow(h,3);	
}

// logaritmo de la funcion de particion de N particulas.
double lnZ (double T, double N, double V) {
	// Usamos la aproximacion de Stirling.
	return (N + N*log(z(V,T)) - N*log(N));
}

// Energia interna del sistema.
double E (double T, double N, double V) {
	double derivate = (lnZ(T+dx, N, V) - lnZ(T, N, V))/dx;
	return k*(pow(T,2))*derivate;
}

// Presion
double p (double T, double N, double V)  {
	double derivate = (z(V+dx, T) - z(V, T))/dx;
	return k*T*N*derivate/z(V,T);
}

void graficarFuncionParticion () {
	std::vector<double> l;
	std::vector<int> temperaturas;
	t1 = clock();
	for (int i = 0; i < Tmaxima; i++) {
		l.push_back(z(1, i+1));
		temperaturas.push_back(i+1);
	}
	t2 = clock();
	double time = (double(t2-t1)/CLOCKS_PER_SEC)/60;
	std::cout << "Tiempo de ejecucion: " << time << " minutos" << std::endl;
	plt::plot(temperaturas,l);
	plt::ylabel("funcion de particion de una particula");
	plt::xlabel("Temperatura");
	plt::title("Espacio-tiempo de Minkowski");
	plt::show();
}

void graficarEvsT () {
	std::vector<double> l;
	std::vector<int> temperaturas;
	t1 = clock();
	for (int i = 0; i < Tmaxima; i++) {
		l.push_back(E(i+1, pow(10,23),1));
		temperaturas.push_back(i+1);
	}
	t2 = clock();
	double time = (double(t2-t1)/CLOCKS_PER_SEC)/60;
	std::cout << "Tiempo de ejecucion: " << time << " minutos" << std::endl;
	plt::plot(temperaturas,l);
	plt::ylabel("Energia");
	plt::xlabel("Temperatura");
	plt::title("Espacio-tiempo de Minkowski");
	plt::show();
}

void graficarPvsT () {
	std::vector<double> l;
	std::vector<int> temperaturas;
	t1 = clock();
	for (int i = 0; i < Tmaxima; i++) {
		l.push_back(p(i+1, pow(10,23),1));
		temperaturas.push_back(i+1);
	}
	t2 = clock();
	double time = (double(t2-t1)/CLOCKS_PER_SEC)/60;
	std::cout << "Tiempo de ejecucion: " << time << " minutos" << std::endl;
	plt::plot(temperaturas,l);
	plt::ylabel("Presion");
	plt::xlabel("Temperatura");
	plt::title("Espacio-tiempo de Minkowski");
	plt::show();
}

int main () {
	graficarFuncionParticion ();
	return 0;
}

File: src/Minkowski/minkowski.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import integrate
from scipy.misc import derivative
from time import time

c = 1
k = 1
m = 1
h = 2*np.pi

# Funcion de particion de una particula.
def z (V,T):
	# Hamiltoniano en relatividad especial.
	H = lambda p_x, p_y, p_z: (np.sqrt((m**2)*(c**4)+(p_x**2 + p_y**2 +p_z**2)*(c**2))-m*c**2)
    	I = integrate.nquad(lambda x, y, z: np.exp(-H(x ,y ,z)/(k*T)), [[-np.inf, np.inf],[-np.inf, np.inf], [-np.inf, np.inf]])
	return I[0]*V*4*np.pi/(h**3)

# logaritmo de la funcion de particion de N particulas.
def lnZ (T, N, V):
	# Usamos la aproximacion de Stirling.
	return (N + N*np.log(z(V,T)) - N*np.log(N))

# Energia interna del sistema.
def E (T, N, V):
	return k*(T**2)*derivative(lambda x: lnZ(x, N, V), T, dx=1e-5)

# Presion
def p (T, N, V):
	return k*T*N/z(V,T)*derivative(lambda x: z(x, T), V, dx=1e-5)

Tmaxima = 100

def graficarEvsT ():
	l = {1: np.array([0]*Tmaxima, dtype=np.float64), 
	     5: np.array([0]*Tmaxima, dtype=np.float64), 
	     10: np.array([0]*Tmaxima, dtype=np.float64)}
	for i in range(0,Tmaxima):
		for key in l:
			l[key][i] = E(i+1, 1e23, key)
	for key in l:
		plt.plot(range(1,Tmaxima+1), l[key], label = "V = {}".format(key))
	plt.legend(loc='upper left')
	plt.ylabel('Energia')
	plt.xlabel('Temperatura')
	plt.title('Espacio-tiempo de Minkowski')
	plt.show()

def graficarPvsT ():
	l = {1: np.array([0]*Tmaxima, dtype=np.float64), 
	     5: np.array([0]*Tmaxima, dtype=np.float64), 
	     10: np.array([0]*Tmaxima, dtype=np.float64)}
    	for i in range(0,Tmaxima):
		for key in l:
			l[key][i] = p(i+1, 1e23, key)
	for key in l:
		plt.plot(range(1,Tmaxima+1), l[key], label = "V = {}".format(key))
    	plt.legend(loc='upper left')
    	plt.ylabel('Presion')
    	plt.xlabel('Temperatura')
    	plt.title('Espacio-tiempo de Minkowski')
    	plt.show()

def graficarFuncionParticion ():
    	l = np.array([0]*Tmaxima, dtype=np.float64)
    	t1 = time()
    	for i in range(0,Tmaxima):
        	l[i] = z(1,i+1)
    	t2 = time()
    	t = (t2 - t1)/60
    	print ("tiempo de ejecucion: "+str(t)+" minutos")
    	plt.plot(range(1,Tmaxima+1), l)
    	plt.xlabel('T')
    	plt.title('funcion de particion de una particula')
    	plt.show()

graficarEvsT()
graficarPvsT()
